### PR TITLE
msgraphforoffice365: Feature - get email eventMessage support

### PR DESCRIPTION
**Pull request by 5hak3y**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [x] The readme.html has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [x] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe): 

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers), 
  please document them in the `readme.html`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `readme.html` all methods (eg, OAuth) used to authenticate 
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `readme.html`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through 
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

  

## Release Notes (REQUIRED)
- Support added for eventMessage in 'get email' action
- Ability to add email message to the vault in 'get email' action

## What is the current behavior? (OPTIONAL)
- Currently it is not possible to retrieve an event associated with a meeting request in a user's mailbox. When deleting an email which is a meeting request it will not delete the event in the user's calendar. 
- Currently is not possible to add an email message to the vault via 'get email' action.

## What is the new behavior? (OPTIONAL)
- The associated event will be retrieved when using 'get email' action on @odata.type of '#microsoft.graph.eventMessage'. This allows the use of action 'delete event' to remove a meeting request from user's calendar.
- If 'download email' parameter in the 'get email' action is enabled the email will be added to the vault. Associated vaultId will be contained in action results.

---
Thanks for contributing!

### CODE REVIEW

**Comment by dhwanis-crest** [comment]
@5hak3y , We have modified the "get_email" action to include "#microsoft.graph.eventMessageRequest" and "#microsoft.graph.eventMessageResponse" type of emails while fetching event-related information. Moreover, in the "delete_event" action we have modified the send_response behavior. 
Can you please check and test these changes from your end? Please let us know if the changes are working as expected or not.

**Comment by dhwanis-crest** [comment]
> @dhwanis-crest Thanks, I've tested the changes and it works as expected.
> 
> I've also added an additional parameter 'download email' to the 'get email' action. This is similar to the 'download attachments', however this adds the complete email message to the vault.

Okay @5hak3y , Please let us know if anything else is remaining from your end or else we can proceed with the testing and release activity for this.

**Comment by dhwanis-crest** [comment]
> @dhwanis-crest Thanks, I've tested the changes and it works as expected.
> 
> I've also added an additional parameter 'download email' to the 'get email' action. This is similar to the 'download attachments', however this adds the complete email message to the vault.

Okay @5hak3y, Please let us know if anything else is remaining from your end, or else we can proceed with the testing and release activity for this.

**Comment by dhwanis-crest** [comment]
The static tests pipeline would pass once the latest playbook MR gets merged.

**Comment by paulnawat** [comment]
> The static tests pipeline would pass once the latest playbook MR gets merged.

In that case, can we merge the playbook MR first? 
This would be consistent with our suggested merge flow as well (i.e. All App dependencies should be merged before the AppChange one)

**Comment by dhwanis-crest** [comment]
@paulnawat, Could you please review this MR? We will merge the code in the next branch only when the static test pipeline passes and all the dependencies are merged in the next branch.

File: office365_connector.py [hunk 1]
```diff
@@ -569,17 +569,22 @@ def _make_rest_call(self, action_result, url, verify=True, headers={}, params=No
...
+                                data=data,
+                                headers=headers,
+                                verify=verify,
+                                params=params,
+                                timeout=MSGOFFICE365_DEFAULT_REQUEST_TIMEOUT)
+            except Exception as e:
+                error_msg = _get_error_message_from_exception(e)
+                return RetVal(action_result.set_status(phantom.APP_ERROR, "Error connecting to server. {0}".format(error_msg)), resp_json)
+            if r.status_code != 502:
+                break
```
Comment (paulnawat) [inline office365_connector.py:585]: should we add debug_log on the response here?

(e.g. `self.debug_print('Received response: %s' % r)` )

File: office365_connector.py [hunk 2]
```diff
@@ -1132,6 +1138,17 @@ def _process_email_data(self, config, action_result, endpoint, email):
 
         return phantom.APP_SUCCESS
 
+    def _remove_tokens(self, action_result):
+        if len(list(filter(lambda x: x in action_result.get_message(), MSGOFFICE365_ASSET_PARAM_CHECK_LIST_ERRORS))) > 0:
```
Comment (paulnawat) [inline office365_connector.py:1142]: nit: should we add a short comment on the long condition?

I believe we are checking if the message includes any of the known error codes.

File: office365_connector.py [hunk 3]
```diff
@@ -1355,6 +1373,30 @@ def _handle_delete_email(self, param):
...
+
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        email_addr = param['email_address']
+        message_id = param['id']
+        send_response = param.get('send_response')
+        endpoint = "/users/{0}".format(email_addr)
+
+        endpoint += '/events/{0}'.format(message_id)
```
Comment (paulnawat) [inline office365_connector.py:1386]: suggestion: `endpoint = "/users/{0}/events/{1}".format(email_addr, message_id)`

File: office365.json [hunk 4]
```diff
@@ -2825,6 +2839,121 @@
...
+                    "description": "Email address of the mailbox owner",
+                    "data_type": "string",
+                    "required": true,
+                    "primary": true,
+                    "contains": [
+                        "email"
+                    ],
+                    "order": 1
+                },
+                "send_response": {
```
Comment (paulnawat) [inline office365.json:2869]: should this be called `send_decline_response`?


`send_response` is quite generic, as response can be anything.

**Review by paulnawat — APPROVED** [review]
lgtm. 

As discussed, please ensure App dependencies (e.g. playbook, unit tests, asset) get merged prior to this PR. Thanks!

File: office365.json (hunk 4 above)
Comment (dhwanis-crest) [reply office365.json]: updated

File: office365_connector.py (hunk 3 above)
Comment (dhwanis-crest) [reply office365_connector.py]: updated

File: office365_connector.py (hunk 2 above)
Comment (dhwanis-crest) [reply office365_connector.py:1143]: updated

File: office365_connector.py (hunk 1 above)
Comment (dhwanis-crest) [reply office365_connector.py:585]: Debugging this way would print the response object and not the content. We have included a debug statement "Received 502 status code from the server" and also regardless of the response content, with 502 status code we have added the retry mechanism

**Review by joshquerinphantom — APPROVED** [review]

**Pull request merged**